Implement Zhuyin parsing and conversion to/from Pinyin

JohnAZoidberg opened the pull request:
Solves #3

JohnAZoidberg committed:
zhuyin: Implement Zhuyin parsing

JohnAZoidberg committed:
pinyin: Parse standalone finals that look like combinations

JohnAZoidberg committed:
zhuyin: Don't allow standalone initials to be combined with final

JohnAZoidberg commented:
Only open point of discussion: https://github.com/JohnAZoidberg/hanyutils/pull/1#discussion_r830147827

JohnAZoidberg commented on lib/hanyutils.ex:
```diff
@@ -59,6 +77,33 @@ defmodule Hanyutils do
+      "ㄋㄧˇ[ ㄏㄠˇ | ㄏㄠˋ ]"
+
+  """
+  # @spec to_zhuyin(String.t(), (Hanzi.t() -> Pinyin.pinyin_list())) :: String.t()
```
Oh, need to uncomment this and fix it.

JohnAZoidberg committed:
zhuyin: Fix dialyzer

JohnAZoidberg commented:
Huh, now I do get some warnings when generating the docs:

```
> mix docs
Compiling 2 files (.ex)
Generating docs...
warning: documentation references function "Hanzi.to_zhuyin/2" but it is undefined or private
  lib/zhuyin.ex:2: Zhuyin

View "html" docs at "doc/index.html"
warning: documentation references function "Hanzi.to_zhuyin/2" but it is undefined or private
  lib/zhuyin.ex:2: Zhuyin

View "epub" docs at "doc/hanyutils.epub"
```

JohnAZoidberg committed:
Clean up and add documentation

mathsaey commented:
Just a heads up: it will take some time before I can look at this properly.

mathsaey commented on README.md:
```diff
@@ -67,9 +70,24 @@ iex> Hanyutils.mark_pinyin("ni3hao3")
+iex> Hanyutils.pinyin_to_zhuyin("ni3hao3")
+"ㄋㄧˇㄏㄠˇ"
+
+iex> Hanyutils.pinyin_to_zhuyin("ni3hǎo")
```
I think there are too many examples here. We can make this a bit shorter by only showing this final example for `pinyin_to_zhuyin`.

JohnAZoidberg replied:
Done

mathsaey commented on lib/hanyutils.ex:
```diff
@@ -32,7 +50,7 @@ defmodule Hanyutils do
       end
 
-  Please refer to the documentation of the `Hanzi` and `Pinyin` modules for more information.
+  Please refer to the documentation of the `Hanzi`, `Pinyin`, and `Zhuyin` modules for more information.
```
As a general note, try to break lines at 98 characters.

JohnAZoidberg replied:
Done

mathsaey commented on lib/hanyutils.ex:
```diff
@@ -59,6 +77,35 @@ defmodule Hanyutils do
+  Convert a string containing Han characters to Zhuyin.
+
+  For more information about `converter`, please refer to `Hanzi.to_pinyin/2`.
+  Because the Unihan database provides only definitions of pinyin
```
Does using the hanzi -> pinyin -> zhuyin "pipeline" have any practical consequences? Otherwise I'd omit this information, since it's effectively just discussing implementation details.

JohnAZoidberg replied:
Similarly as `to_marked_pinyin` and `to_numbered_pinyin` there's a `converter` parameter that allows choosing the pronounciations and defaults to `Hanzi.common_pronunciation`.
So yes, because we get the pronounciations from the Hanzi via Pinyin and it's configurable, the consumer should know.

mathsaey commented on lib/hanyutils.ex:
```diff
@@ -120,4 +167,78 @@ defmodule Hanyutils do
   end
+
+  @doc """
+  Convert a string with marked Pinyin to Zhuyin
```
I'd be careful with phrasing here. Saying "marked" makes me think this function only accepts marked pinyin.

mathsaey commented on lib/hanyutils.ex:
```diff
@@ -120,4 +167,78 @@ defmodule Hanyutils do
+  @doc """
+  Convert a string with marked Pinyin to Zhuyin
+
+  Parses the input using `Pinyin.read!/1` (in `:words` mode), and converts the result with
```
Perhaps we can add the mode as a second argument and use `:words` as the default value for this argument? That would make this function a bit more flexible.

JohnAZoidberg replied:
Sure, done. Also for the pinyin functions.

mathsaey commented on lib/hanyutils.ex:
```diff
@@ -120,4 +167,78 @@ defmodule Hanyutils do
+  Convert a string with marked Pinyin to Zhuyin
+
+  Parses the input using `Pinyin.read!/1` (in `:words` mode), and converts the result with
+  `Pinyin.numbered/1`. Please refer to the documentation of `Pinyin.read/2` if you required
```
required -> require

JohnAZoidberg replied:
Done

mathsaey commented on lib/hanyutils.ex:
```diff
@@ -120,4 +167,78 @@ defmodule Hanyutils do
+  @doc """
+  Convert a string with Zhuyin to marked Pinyin.
+
+  Parses the input using `Zhuyin.read!/1` (in `:words` mode), and converts the result with
```
Same remark about the mode here.

mathsaey commented on lib/hanyutils.ex:
```diff
@@ -120,4 +167,78 @@ defmodule Hanyutils do
+  @doc """
+  Convert a string with Zhuyin to numbered Pinyin.
+
+  Parses the input using `Zhuyin.read!/1` (in `:words` mode), and converts the result with
```
Same remark about the mode here.

mathsaey commented on lib/hanyutils.ex:
```diff
@@ -120,4 +167,78 @@ defmodule Hanyutils do
+
+  """
+  @spec zhuyin_to_marked(String.t()) :: String.t()
+  def zhuyin_to_marked(string) do
```
I'm thinking if we should name this `zhuyin_to_marked_pinyin` to be explicit here. Any thoughts?

mathsaey commented on lib/hanyutils.ex:
```diff
@@ -120,4 +167,78 @@ defmodule Hanyutils do
+      "zhu1yu3chen2"
+
+  """
+  @spec zhuyin_to_numbered(String.t()) :: String.t()
```
If we rename `zhuyin_to_marked` to `zhuyin_to_marked_pinyin`, we should do the same here of course.

JohnAZoidberg replied:
Sure. Might be confusing for people who don't know Zhuyin, for example that it doesn't have a numbered version.
The only issue is that the function name would be quite long.

mathsaey commented on lib/pinyin/parsers.ex:
```diff
@@ -58,6 +58,14 @@ defmodule Pinyin.Parsers do
+  # Standalone finals that look like initial and final
+  # Must be parsed as final, otherwise conversion to zhuyin won't be correct.
+  # For example 吃 is chi where the i is not a final.
+  surprising_standalone_finals = ~w(
```
This is ugly, but I guess it cannot be avoided?

If I understand it correctly, if "i" is used as a final with these initials it always counts as a "complete" final?

JohnAZoidberg replied:
These finals here behave very similarly as the other standalone finals. The difference is that the entries in the already existing list of standalone finals are unique. For example "wang" - if we didn't have the standalone finals list, it wouldn't be able to be parsed. So it is always a standalone final, never a combination.
But "chi" is a standalone final by meaning but the parser would incorrectly classify it as an initial ch+i.

It's not just an artifact of conversion to Zhuyin. If you try to pronounce "chi" you see that the "i" is not a final. It's not pronounced like the "i" sound (which confusingly is spelled as "yi" when used as a standalone final). Where the "i" in "qi" has the "i"/"yi" sound.

That's one reason why I like to know Zhuyin and Pinyin. Makes me understand better how the sounds are composed.

mathsaey replied:
Okay. Perhaps we could also refer to the bopomofo wikipedia page here so that we have some reference we can look at if we ever need to adjust the parser again.

Besides that, let's hope we don't find some uncommon character in a year or so where "chi" _is_ an initial + final.

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+  """
+  @type zhuyin_list :: [t() | String.t()]
+
+  def zhuyin_tones, do: ["˙", "", "ˊ", "ˇ", "ˋ"]
```
Any reason this is public?

If you use this in another module, add `@doc false`, and start the name with a `_`, as noted [here](https://hexdocs.pm/elixir/1.13/writing-documentation.html#hiding-internal-modules-and-functions)

JohnAZoidberg replied:
It's used down below in the implementation of `defimpl String.Chars, for: Zhuyin do`.
Seems it has to be public for that. Or is there a better way to do it?

JohnAZoidberg replied:
That's also why it's a funciton and not a variable. Not an Elixir expert, as I said, so it might be stupid.

JohnAZoidberg replied:
Damn, if I generate `tone_index` at compile-time is can only access variables (not sure if that's the right name) but not the module function. I assume because at that point in time the module isn't available yet.

mathsaey replied:
>  Seems it has to be public for that. Or is there a better way to do it?

Nope, you can't avoid that. In that case, just mark it as `@doc false` and start the name with an underscore :).

> Damn, if I generate tone_index at compile-time is can only access variables (not sure if that's the right name) but not the module function. I assume because at that point in time the module isn't available yet.

Correct. However, you can generate multiple functions based on compile-time variables. E.g. instead of writing this:

```elixir
  def zhuyin_tones, do: ["˙", "", "ˊ", "ˇ", "ˋ"]
  zhuyin_tones_var = ["˙", "", "ˊ", "ˇ", "ˋ"]

  for {marker, idx} <- Enum.with_index(zhuyin_tones_var, 0) do
    defp tone_index(unquote(marker)), do: unquote(idx)
  end
```

you could instead write:

```elixir
  zhuyin_tones = ["˙", "", "ˊ", "ˇ", "ˋ"]
  def zhuyin_tones, do: unquote(zhuyin_tones)

  for {marker, idx} <- Enum.with_index(zhuyin_tones_var, 0) do
    defp tone_index(unquote(marker)), do: unquote(idx)
  end
```

I think that should work.

JohnAZoidberg replied:
it does work, thanks!

Do you know why `unquote` is required here?

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+
+  def zhuyin_tones, do: ["˙", "", "ˊ", "ˇ", "ˋ"]
+
+  defp tone_index(tone_char) do
```
We could generate this function at compile time, similar to how `mark` is generated in the `Pinyin` module. Probably won't make that much of a difference, but otherwise we'd need to traverse this list every time.

JohnAZoidberg replied:
Done.

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+    %Pinyin{initial: "", final: "r", tone: "0"}
+  end
+
+  def to_pinyin(zhuyin = %Zhuyin{}) do
```
I'd write this as:

```
if initial = @initials[zhuyin.initial] do
  %Pinyin{initial: intial, final: @finals[zhuyin.final], tone: zhuyin.tone}
else
  %Pinyin{initial: "", final: @standalone_finals[zhuyin.final], tone: zhuyin.tone}
end
```

It's a bit more concise and (in my opinion) idiomatic.

JohnAZoidberg replied:
That doesn't seem to work:

```
> mix test
Compiling 4 files (.ex)
warning: variable "intial" does not exist and is being expanded to "intial()", please use parentheses to remove the ambiguity or change the variable name
  lib/zhuyin.ex:178: Zhuyin.to_pinyin/1

warning: variable "initial" is unused (if the variable is not meant to be used, prefix it with an underscore)
  lib/zhuyin.ex:177: Zhuyin.to_pinyin/1


== Compilation error in file lib/zhuyin.ex ==
** (CompileError) lib/zhuyin.ex:178: undefined function intial/0 (expected Zhuyin to define such a function or for it to be imported, but none are available)
```

![image](https://user-images.githubusercontent.com/5307138/164466090-2c0b7103-76e2-4291-a7a0-644deb257fb6.png)

mathsaey replied:
You made a typo: `intial` instead of `initial` (on line 178).

JohnAZoidberg replied:
Oh.. Right now it works.

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+  @spec to_pinyin(t() | zhuyin_list()) :: Pinyin.t() | Pinyin.pinyin_list()
+  # Special case for this final and tone combination
+  def to_pinyin(%Zhuyin{initial: "", final: "ㄦ", tone: 0}) do
+    %Pinyin{initial: "", final: "r", tone: "0"}
```
This should be `final: "er"`, to be consistent with the result of `Hanzi.from_character("儿").pron.final`.

(Handling 这儿 etc is something I want to improve later, but I think it is important that the pinyin of the "standalone" character remains correct).

JohnAZoidberg replied:
Ok, makes sense to have pinyin and zhuyin agree.

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+  @doc """
+  Create a pinyin struct from a zhuyin struct or list.
+
+  ## Examples
```
Please use the representation shown in iex (`#Zhuyin<...>`) here to be as close as possible to what the user sees in the shell.

JohnAZoidberg replied:
I initially used that but it won't work because it's two syllables and looks like this in iex:

```
iex> Zhuyin.from_pinyin(~p/nǐhǎo/)
[#Zhuyin<ㄋㄧˇ>, #Zhuyin<ㄏㄠˇ>]
```

however, mix test has issues parsing that because `#` looks like a comment:

```
  1) doctest Zhuyin.from_pinyin/1 (1) (ZhuyinTest)
     test/zhuyin_test.exs:5
     Doctest did not compile, got: (TokenMissingError) lib/zhuyin.ex:192:2: missing terminator: ] (for "[" starting at line 192)
          |
      192 | [#Zhuyin<ㄋㄧˇ>, #Zhuyin<ㄏㄠˇ>]
          |  ^
     If you are planning to assert on the result of an iex> expression which contains a value inspected as #Name<...>, please make sure the inspected value is placed at the beginning of the expression; otherwise Elixir will treat it as a comment due to the leading sign #.
     doctest:
       iex> Zhuyin.from_pinyin(~p/nǐhǎo/)
       [#Zhuyin<ㄋㄧˇ>, #Zhuyin<ㄏㄠˇ>]
     stacktrace:
       lib/zhuyin.ex:192: Zhuyin (module)
```

What do you think? Should I reduce the example to a single syllable to be able to use the `#Zhuyin<...>` notation?

mathsaey replied:
Right, I forgot `from_pinyin` returns a list. The matching only works if the line starts with `#` (https://hexdocs.pm/ex_unit/1.13/ExUnit.DocTest.html#module-opaque-types). In that case I'd use the full list representation (like how the doctests for `Pinyin.read` are written).

JohnAZoidberg replied:
Okay, expanded it. It's also better because it doesn't involve another parsing and makes the test more pure.

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+  """
+  @spec from_pinyin(Pinyin.t() | Pinyin.pinyin_list()) :: t()
+  def from_pinyin(pinyin = %Pinyin{}) do
+    initial_map = Map.new(@initials, fn {key, val} -> {val, key} end)
```
This will create a new map every time this function is called. Perhaps you can create this map at compile time (`@foo Map.new(@initials, fn {key, val} -> {val, key} end)`) to avoid this.

JohnAZoidberg replied:
Good idea! Done.

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+    initial = Map.get(initial_map, pinyin.initial, "")
+
+    final =
+      if pinyin.initial == "" do
```
I'd rework this if similar to how I proposed you rework `to_pinyin`.

JohnAZoidberg replied:
Done

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+
+    final =
+      if pinyin.initial == "" do
+        standalone_finals_map = Map.new(@standalone_finals, fn {key, val} -> {val, key} end)
```
Same remark as before, please generate this map at compile time.

JohnAZoidberg replied:
Done.

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+        standalone_finals_map = Map.new(@standalone_finals, fn {key, val} -> {val, key} end)
+        standalone_finals_map[pinyin.final]
+      else
+        finals_map = Map.new(@finals, fn {key, val} -> {val, key} end)
```
Same remark as before, please generate this map at compile time.

JohnAZoidberg replied:
Done

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+    end)
+  end
+
+  def create(initial, final, tone) do
```
I think this function should be private or, if it is not used, removed, so that the zhuyin and pinyin modules are more consistent with each other.

mathsaey commented on lib/zhuyin.ex:
```diff
@@ -0,0 +1,376 @@
+  defp parser_result!({:error, _, rem, %{}, _, _}), do: raise(ParseError, rem)
+
+  @doc """
+  Create a zhuyin struct (`t:t/0`) from a string.
```
I would use the word "single" here, to make this explicit in the summary generated at the top of the docs.

mathsaey replied:
I should probably do the same in the `Pinyin` module :).

JohnAZoidberg replied:
What do you mean "at the top of the docs"? At the top of the docs for that function?

mathsaey replied:
Yes, sorry for not being more clear. The first line is shown in the summary so I think the extra context is useful there. The summary is shown at the top of the module's docpage, hence why I said "top of the docs".

JohnAZoidberg replied:
Oh I see. Makes sense.

mathsaey reviewed (changes requested):
Sorry this took me so long. I think this looks good overall, but I do have a bunch of remarks I'd like to address.

JohnAZoidberg committed:
Fix up documentation

JohnAZoidberg committed:
Hanyutils: Add _to_pinyin to make it explicit

Might be confusing that it converts to pinyin for people who don't know
that zhuyin only has a marked variant.

JohnAZoidberg commented:
> Sorry this took me so long. I think this looks good overall, but I do have a bunch of remarks I'd like to address.

No problem, I'm not shipping a product built on this :D
Thanks for taking the time to review!

JohnAZoidberg committed:
Zhuyin: Generate tone_index at compile-time

JohnAZoidberg committed:
Zhuyin: Reverse maps at compile-time

Otherwise it'll be done every time from_pinyin is called.

JohnAZoidberg committed:
Hanyutils: Allow custom parsing mode

JohnAZoidberg committed:
Zhuyin: Deduplicate zhuyin_tones

JohnAZoidberg committed:
Zhuyin: Simplify to_/from_pinyin

JohnAZoidberg committed:
Zhuyin: Expand example

Make it more like what it would actually look like in iex.

JohnAZoidberg committed:
Zhuyin: Remove special case for ㄦ

Pinyin also doesn't have it. We should have both in sync.

mathsaey committed:
Documentation updates

mathsaey committed:
Fix code example in readme

mathsaey committed:
Documentation updates

mathsaey committed:
Adjust surprising_standalone finals name and documentation

mathsaey commented on test/zhuyin_test.exs:
```diff
@@ -0,0 +1,112 @@
+    end
+
+    test "Works for syllable with initial, final and tone" do
+      {:ok, [%Zhuyin{initial: "ㄋ", final: "ㄧ", tone: 3}], "", %{}, {1, 0}, 8} =
```
You should probably put this in an assertion using `==` instead. I would not rely on the match error here. Alternatively, you could use the `match?` function.

No need to fix this, I'm working on this file, but thought I'd point it out for future reference.

mathsaey commented on test/zhuyin_test.exs:
```diff
@@ -0,0 +1,112 @@
+  end
+
+  describe "Syllable parser" do
+    test "Syllable parser parses only a single syllable" do
```
ExUnit tests are shown as a single line, e.g.: "Syllable parser Syllable parser parses only a single syllable". So I personally only use lowercase names and try to turn the text into a logical sentence. e.g. in this case you could turn the describe into `"from_string"` and the test names into something like "parses only a single syllable".

Once again, no need to modify this, I'm going through the code myself at the moment.

mathsaey commented on test/zhuyin_test.exs:
```diff
@@ -0,0 +1,112 @@
+
+      # That can only be parsed if we allow multiple words
+      assert {:ok, [~z/ㄕ/, ~z/ㄧˇ/] |> List.flatten(), "", %{}, {1, 0}, 8} ==
+               Zhuyin.Parsers.zhuyin_only("ㄕㄧˇ")
```
I'm changing the tests to use `Zhuyin.read` and others instead of the parsers, as I consider these to be an implementation detail.

mathsaey committed:
Reorganize tests

mathsaey committed:
Remove unnecessary as: from alias

mathsaey commented on lib/zhuyin/parsers.ex:
```diff
@@ -0,0 +1,99 @@
+  alias Pinyin.Parsers.Wordlist, as: Wordlist
+
+  # Empty tone must be last for parsing to work
+  zhuyin_tones = ["˙", "ˊ", "ˇ", "ˋ", ""]
```
I feel it is a bit strange to add the empty string as the final tone here. Would it not make more sense to just assume no tone (as the tone parser is optional anyway), like in the pinyin parser?

mathsaey committed:
change formatting

mathsaey committed:
Documentation formatting

mathsaey committed:
Remove optional from tone parser as the empty string always matches

mathsaey committed:
Fail when final is not present in @finals or in @standalone finals

Seems to be preferrable over returning an invalid result.

mathsaey committed:
Move zhuyin creation to parser module

mathsaey committed:
Create _tone_for_index function, make _zhuyin_tones private

mathsaey committed:
Make create private

mathsaey committed:
Uncomment + fix doctests

mathsaey committed:
Various cosmetic changes

- use parser_result in read
- various whitespace changes
- remove typespec from to_charlist

mathsaey committed:
Run mix format

mathsaey commented:
@JohnAZoidberg: I went through most of the files and this looks okay to me now. Could you just add the documentation for `Zhuyin.read` and `Zhuyin.read!`?

If you have the time, perhaps it might also be a good idea to go through my commits (or the current state of things) and see if you agree with the changes I made.

JohnAZoidberg commented:
Ok, thanks. I'll have a look soon

mathsaey committed:
Add documentation to read/2, read!/2

mathsaey commented:
I have added documentation to `read/2` and `read!/2`.
Unless you have any remarks @JohnAZoidberg, I'll close and merge this soon :).
